spring-statemachine-core/src/main/java/org/springframework/statemachine/config/StateMachineBuilder.java

```diff
@@ -112,40 +109,37 @@ public Builder() {
 		}
 
 		/**
-		 * Builds a {@link StateMachine}.
+		 * Creates a {@link StateMachineFactory} from builder
 		 *
-		 * @return the state machine
+		 * @return the factory to create a state machine
 		 */
-		public StateMachine<S, E> build() {
+		public StateMachineFactory<S, E> createFactory() {
 			try {
 				builder.apply(adapter);
-				StateMachineConfig<S, E> stateMachineConfig = builder.getOrBuild();
-
-				TransitionsData<S, E> stateMachineTransitions = stateMachineConfig.getTransitions();
-				StatesData<S, E> stateMachineStates = stateMachineConfig.getStates();
-				ConfigurationData<S, E> stateMachineConfigurationConfig = stateMachineConfig.getStateMachineConfigurationConfig();
-
-				ObjectStateMachineFactory<S, E> stateMachineFactory = null;
-				if (stateMachineConfig.getModel() != null && stateMachineConfig.getModel().getFactory() != null) {
-					stateMachineFactory = new ObjectStateMachineFactory<S, E>(
-							new DefaultStateMachineModel<S, E>(stateMachineConfigurationConfig, null, null),
-							stateMachineConfig.getModel().getFactory());
-				} else {
-					stateMachineFactory = new ObjectStateMachineFactory<S, E>(new DefaultStateMachineModel<S, E>(
-							stateMachineConfigurationConfig, stateMachineStates, stateMachineTransitions), null);
-				}
+
+				ObjectStateMachineFactory<S, E> stateMachineFactory = StateMachineFactory.create(builder);
+				ConfigurationData<S, E> stateMachineConfigurationConfig = builder.getOrBuild().stateMachineConfigurationConfig;
 
 				stateMachineFactory.setHandleAutostartup(stateMachineConfigurationConfig.isAutoStart());
 
 				if (stateMachineConfigurationConfig.getBeanFactory() != null) {
 					stateMachineFactory.setBeanFactory(stateMachineConfigurationConfig.getBeanFactory());
 				}
-				return stateMachineFactory.getStateMachine();
+				return stateMachineFactory;
```

Comment:
I ran tests(rebased) and `ConfigurationErrorTests` do fail as `MalformedConfigurationException` is not wrapped with `StateMachineException`. I think same try/catch should be in a new `build()` method as well.